extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -0,0 +1,685 @@
+              bodyUtf16CharOffsets: !Array<number>,
+              bodyAttributes: ?string,
+              customElementExtensions: Array<string>,
+              customStylesheets: }} */
+let CreativeMetaDataDef;
+
+export class AmpA4A extends AMP.BaseElement {
+  // TODO: Add more error handling throughout code.
+  // TODO: Handle creatives that do not fill.
+
+  /**
+   * @param {!Element} element
+   */
+  constructor(element) {
+    super(element);
+
+    /** @private {?Promise<!boolean>} */
+    this.adPromise_ = null;
+
+    /**
+     * @private {number} unique ID of the currently executing promise to allow
+     *    for cancellation.
+     */
+    this.promiseId_ = 0;
+
+    /** @private {?string} */
+    this.adUrl_ = null;
+
+    /** @private {boolean} */
+    this.rendered_ = false;
+
+    /** @private {number} ID of timer used as part of 3p throttling. */
+    this.timerId_ = 0;
+
+    /** @private {null|boolean} where layoutMeasure has been executed. */
+    this.layoutMeasureExecuted_ = false;
+
+    /**
+     * @private {!Array<!Element>} stylesheets added as part of shadow DOM
+     *    based creative injection.
+     */
+    this.stylesheets_ = [];
+
+    /** @param {IntersectionObserver} */
+    this.intersectionObserver_ = null;
+  }
+
+  /** @override */
+  getPriority() {
+    // Priority used for scheduling preload and layout callback.  Because
+    // AMP creatives will be injected as part of the promise chain created
+    // within onLayoutMeasure, this is only relevant to non-AMP creatives
+    // therefore we want this to match the 3p priority.
+    return 2;
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /** @override */
+  renderOutsideViewport() {
+    // Only relevant if non-AMP as AMP creative will be injected within
+    // buildCallback promise chain.
+    // If another ad is currently loading we only load ads that are currently
+    // in viewport.
+    const allowRender = allowRenderOutsideViewport(this.element, this.getWin());
+    if (allowRender !== true) {
+      return allowRender;
+    }
+    // Otherwise the ad is good to go.
+    return super.renderOutsideViewport();
+  }
+
+  /**
+   * To be overridden by network specific implementation indicating if element
+   * (and environment generally) are valid for sending XHR rqueries.
+   * @return {boolean} where element is valid and ad request should be sent.  If
+   *    false, no ad request is sent and slot will be collapsed if possible.
+   */
+  isValidElement() {
+    return true;
+  }
+
+  /**
+   * Prefetches and preconnects URLs related to the ad using adPreconnect
+   * registration which assumes ad request domain used for 3p is applicable.
+   * @override
+   */
+  preconnectCallback(unusedOnLayout) {
+    const preconnect = adPreconnect[this.element.getAttribute('type')];
+    // NOTE(keithwrightbos): using onLayout to indicate if preconnect should be
```

Comment:
I don't understand the comment. The arg doesn't delay preconnect, just influences when the next is made. It should be set to true if the request chain is invoked through actual layout.